### PR TITLE
[Enhancement] prune duplicated columns in partition expr

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [x] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Refers to: https://github.com/StarRocks/starrocks/pull/7486

To fix following plan. As you can see there are two equal join conjuncts, and one side are duplicated columns.

And fragment 6 output are shuffled by parttion expr `1: p_partkey, 1: p_partkey `. It can be optimized to `1:p_partkey`

```
|   27:HASH JOIN                                                                                                                                          |
|   |  join op: INNER JOIN (PARTITIONED)                                                                                                                  |
|   |  equal join conjunct: [17: ps_partkey, INT, true] = [1: p_partkey, INT, true]                                                                       |
|   |  equal join conjunct: [29: ps_partkey, INT, true] = [1: p_partkey, INT, true]                                                                       |
|   |  output columns: 1, 3, 18                                                                                                                           |
|   |  cardinality: 80000000                                                                                                                              |
|   |  column statistics:                                                                                                                                 |
|   |  * p_partkey-->[-Infinity, Infinity, 0.0, 1.0, 1.0] UNKNOWN                                                                                         |
|   |  * p_mfgr-->[-Infinity, Infinity, 0.0, 1.0, 1.0] UNKNOWN                                                                                            |
|   |  * ps_partkey-->[-Infinity, Infinity, 0.0, 1.0, 1.0] UNKNOWN                                                                                        |
|   |  * ps_suppkey-->[-Infinity, Infinity, 0.0, 1.0, 1.0] UNKNOWN                                                                                        |
|   |  * ps_partkey-->[-Infinity, Infinity, 0.0, 1.0, 1.0] UNKNOWN                                                                                        |
|   |                                                                                                                                                     |
|   |----26:EXCHANGE                                                                                                                                      |
|   |       cardinality: 2500000                                                                                                                          |
|   |                                                                                                                                                     |
|   23:EXCHANGE                                                                                                                                           |
|      cardinality: 80000000                                                                                                                              |
|                                                                                                                                                         |
| PLAN FRAGMENT 6(F15)                                                                                                                                    |
|                                                                                                                                                         |
|   Input Partition: RANDOM                                                                                                                               |
|   OutPut Partition: HASH_PARTITIONED: 1: p_partkey, 1: p_partkey                                                                                        |
|   OutPut Exchange Id: 26                                                                                                                                |
|                                                                                                                                                         |
```

And after pruning, we get following plan, bonus with a global runtime filter

```
|   26:HASH JOIN                                                                                                                                          |
|   |  join op: INNER JOIN (PARTITIONED)                                                                                                                  |
|   |  equal join conjunct: [17: ps_partkey, INT, true] = [1: p_partkey, INT, true]                                                                       |
|   |  equal join conjunct: [29: ps_partkey, INT, true] = [1: p_partkey, INT, true]                                                                       |
|   |  build runtime filters:                                                                                                                             |
|   |  - filter_id = 4, build_expr = (1: p_partkey), remote = true                                                                                        |
|   |  output columns: 1, 3, 18                                                                                                                           |
|   |  cardinality: 64800000                                                                                                                              |
|   |  column statistics:                                                                                                                                 |
|   |  * p_partkey-->[-Infinity, Infinity, 0.0, 1.0, 1.0] UNKNOWN                                                                                         |
|   |  * p_mfgr-->[-Infinity, Infinity, 0.0, 1.0, 1.0] UNKNOWN                                                                                            |
|   |  * ps_partkey-->[-Infinity, Infinity, 0.0, 1.0, 1.0] UNKNOWN                                                                                        |
|   |  * ps_suppkey-->[-Infinity, Infinity, 0.0, 1.0, 1.0] UNKNOWN                                                                                        |
|   |  * ps_partkey-->[-Infinity, Infinity, 0.0, 1.0, 1.0] UNKNOWN                                                                                        |
|   |                                                                                                                                                     |
|   |----25:EXCHANGE                                                                                                                                      |
|   |       cardinality: 2500000                                                                                                                          |
|   |                                                                                                                                                     |
|   22:EXCHANGE                                                                                                                                           |
|      cardinality: 64800000                                                                                                                              |
|                                                                                                                                                         |
| PLAN FRAGMENT 6(F13)                                                                                                                                    |
|                                                                                                                                                         |
|   Input Partition: RANDOM                                                                                                                               |
|   OutPut Partition: HASH_PARTITIONED: 1: p_partkey                                                                                                      |
|   OutPut Exchange Id: 25                                                                                                                                |
```
```

### CODE REVIEW

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/physical/PruneShuffleColumnRule.java|8|8|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/physical/PruneShuffleColumnRule.java [hunk 1]
```diff
@@ -141,6 +152,10 @@ private void prune(DistributionContext childContext) {
                 }
             }
 
+            if (duplicatedColumns && maxColumnIndex == -1) {
+                maxColumnIndex = 0;
+            }
+
```
Comment (Seaven) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/physical/PruneShuffleColumnRule.java:158]: you will choose a bad column when not exists statistic, or choose a low cardinality column

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 8 / 8 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rewrite/physical/PruneShuffleColumnRule.java|8|8|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/physical/PruneShuffleColumnRule.java [hunk 2]
```diff
@@ -112,6 +113,16 @@ private void prune(DistributionContext childContext) {
                 return;
             }
 
+            // check if any one of them are duplicated columns like [1,1] = [17,29]
```
Comment (Youngwb) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/physical/PruneShuffleColumnRule.java:116]: Like [a,b,b] = [c,d,e], I think we also could prune the columns

Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rewrite/physical/PruneShuffleColumnRule.java:116]: in theory yes, but there is degraded case.

**Comment by dirtysalt** [comment]
@Youngwb I have not thought clearly about this PR.

there might be degraded case. For example `[a,a] = [c,d]`

assume a has 10 ndv, c has 1 ndv, d has 20 ndv. if we select `c` as partition expression, there is a hot partition(we get 1 ndv). but if we use `[c,d]` as partition expression, we get 20 ndv.

**Pull request closed without merging**